## The Strategy 

Check the past 12 month performance of all Nifty stocks
Buy the top performers and hold them for 1 month.

In [2]:
import yfinance as yf
import numpy as np
import pandas as pd

In [7]:
stocks=pd.read_html('https://en.wikipedia.org/wiki/NIFTY_50')[2].Symbol
stocks.head()

0      ADANIENT
1    ADANIPORTS
2    APOLLOHOSP
3    ASIANPAINT
4      AXISBANK
Name: Symbol, dtype: object

In [8]:
stocks= stocks +'.NS' #yfincane need to add this to non US stocks

In [9]:
stocks=stocks.to_list()

In [11]:
df=yf.download(stocks,start='2012-01-01')['Close']
df

[*********************100%%**********************]  50 of 50 completed


ADANIENT.NS  ADANIPORTS.NS  APOLLOHOSP.NS  ASIANPAINT.NS  \
Date                                                                   
2012-01-02    44.242844     114.699997     552.599976     259.029999   
2012-01-03    44.043587     120.300003     557.150024     265.230011   
2012-01-04    42.633457     122.849998     568.200012     267.494995   
2012-01-05    40.939770     126.699997     571.400024     267.170013   
2012-01-06    41.162018     127.699997     570.900024     266.059998   
...                 ...            ...            ...            ...   
2024-05-22  3140.949951    1378.199951    5868.149902    2885.750000   
2024-05-23  3387.300049    1443.349976    5970.049805    2903.750000   
2024-05-24  3384.949951    1416.099976    5914.350098    2874.750000   
2024-05-27  3289.050049    1431.650024    5939.299805    2872.300049   
2024-05-28  3244.050049    1400.500000    5904.500000    2911.199951   

            AXISBANK.NS  BAJAJ-AUTO.NS  BAJAJFINSV.NS  BAJFINANCE.NS  \
Date                                                                   
2012-01-02   158.960007    1475.500000      41.268360      58.625214   
2012-01-03   167.509995    1494.500000      42.855038      59.553074   
2012-01-04   169.630005    1425.550049      41.096428      58.809811   
2012-01-05   174.130005    1468.599976      41.317482      58.979839   
2012-01-06   170.699997    1456.750000      41.440289      57.323292   
...                 ...            ...            ...            ...   
2024-05-22  1126.000000    8805.549805    1583.949951    6744.549805   
2024-05-23  1165.250000    8963.450195    1608.449951    6826.649902   
2024-05-24  1174.000000    8948.200195    1601.699951    6836.899902   
2024-05-27  1187.000000    8998.700195    1595.800049    6895.500000   
2024-05-28  1183.400024    8989.750000    1599.650024    6873.049805   

            BHARTIARTL.NS     BPCL.NS  ...  SHRIRAMFIN.NS  SUNPHARMA.NS  \
Date                                   ...                                
2012-01-02     310.753510   80.083336  ...     409.228302    247.600006   
2012-01-03     323.235931   78.791664  ...     426.912811    249.725006   
2012-01-04     312.150452   78.449997  ...     435.584534    250.625000   
2012-01-05     310.393005   76.650002  ...     477.043274    249.925003   
2012-01-06     297.910583   79.416664  ...     468.420227    249.975006   
...                   ...         ...  ...            ...           ...   
2024-05-22    1347.900024  640.500000  ...    2336.750000   1539.300049   
2024-05-23    1374.050049  647.450012  ...    2400.449951   1495.099976   
2024-05-24    1388.500000  654.099976  ...    2397.500000   1486.699951   
2024-05-27    1384.349976  656.049988  ...    2380.000000   1466.050049   
2024-05-28    1371.050049  646.500000  ...    2393.699951   1464.900024   

            TATACONSUM.NS  TATAMOTORS.NS  TATASTEEL.NS       TCS.NS  \
Date                                                                  
2012-01-02      90.949997     181.997894     32.497845   589.299988   
2012-01-03      94.550003     191.891769     34.474773   598.799988   
2012-01-04      95.400002     199.064835     34.674850   586.500000   
2012-01-05      93.099998     199.955276     34.546230   586.000000   
2012-01-06      93.500000     201.439362     34.646267   584.700012   
...                   ...            ...           ...          ...   
2024-05-22    1120.349976     947.500000    173.300003  3832.000000   
2024-05-23    1117.949951     962.349976    175.500000  3893.449951   
2024-05-24    1098.250000     960.549988    174.850006  3849.500000   
2024-05-27    1087.900024     958.500000    175.500000  3847.050049   
2024-05-28    1095.550049     947.549988    174.899994  3839.899902   

               TECHM.NS     TITAN.NS  ULTRACEMCO.NS    WIPRO.NS  
Date                                                             
2012-01-02   142.975006   175.149994    1154.000000  149.625031  
2012-01-03   149.987503   179.550003    

In [12]:
ret_df=df.pct_change()
ret_df

ADANIENT.NS  ADANIPORTS.NS  APOLLOHOSP.NS  ASIANPAINT.NS  \
Date                                                                   
2012-01-02          NaN            NaN            NaN            NaN   
2012-01-03    -0.004504       0.048823       0.008234       0.023935   
2012-01-04    -0.032017       0.021197       0.019833       0.008540   
2012-01-05    -0.039727       0.031339       0.005632      -0.001215   
2012-01-06     0.005429       0.007893      -0.000875      -0.004155   
...                 ...            ...            ...            ...   
2024-05-22     0.007522      -0.005269      -0.013466       0.011905   
2024-05-23     0.078432       0.047272       0.017365       0.006238   
2024-05-24    -0.000694      -0.018880      -0.009330      -0.009987   
2024-05-27    -0.028331       0.010981       0.004219      -0.000852   
2024-05-28    -0.013682      -0.021758      -0.005859       0.013543   

            AXISBANK.NS  BAJAJ-AUTO.NS  BAJAJFINSV.NS  BAJFINANCE.NS  \
Date                                                                   
2012-01-02          NaN            NaN            NaN            NaN   
2012-01-03     0.053787       0.012877       0.038448       0.015827   
2012-01-04     0.012656      -0.046136      -0.041036      -0.012481   
2012-01-05     0.026528       0.030199       0.005379       0.002891   
2012-01-06    -0.019698      -0.008069       0.002972      -0.028087   
...                 ...            ...            ...            ...   
2024-05-22    -0.010240      -0.001718       0.000885       0.000252   
2024-05-23     0.034858       0.017932       0.015468       0.012173   
2024-05-24     0.007509      -0.001701      -0.004197       0.001501   
2024-05-27     0.011073       0.005644      -0.003684       0.008571   
2024-05-28    -0.003033      -0.000995       0.002413      -0.003256   

            BHARTIARTL.NS   BPCL.NS  ...  SHRIRAMFIN.NS  SUNPHARMA.NS  \
Date                                 ...                                
2012-01-02            NaN       NaN  ...            NaN           NaN   
2012-01-03       0.040168 -0.016129  ...       0.043214      0.008582   
2012-01-04      -0.034295 -0.004336  ...       0.020313      0.003604   
2012-01-05      -0.005630 -0.022944  ...       0.095180     -0.002793   
2012-01-06      -0.040215  0.036095  ...      -0.018076      0.000200   
...                   ...       ...  ...            ...           ...   
2024-05-22       0.002678 -0.005435  ...      -0.015380     -0.001071   
2024-05-23       0.019401  0.010851  ...       0.027260     -0.028714   
2024-05-24       0.010516  0.010271  ...      -0.001229     -0.005618   
2024-05-27      -0.002989  0.002981  ...      -0.007299     -0.013890   
2024-05-28      -0.009607 -0.014557  ...       0.005756     -0.000784   

            TATACONSUM.NS  TATAMOTORS.NS  TATASTEEL.NS    TCS.NS  TECHM.NS  \
Date                                                                         
2012-01-02            NaN            NaN           NaN       NaN       NaN   
2012-01-03       0.039582       0.054363      0.060833  0.016121  0.049047   
2012-01-04       0.008990       0.037381      0.005804 -0.020541  0.007417   
2012-01-05      -0.024109       0.004473     -0.003709 -0.000853 -0.014974   
2012-01-06       0.004296       0.007422      0.002896 -0.002218  0.006047   
...                   ...            ...           ...       ...       ...   
2024-05-22       0.024133      -0.003995     -0.006022  0.003089  0.004910   
2024-05-23      -0.002142       0.015673      0.012695  0.016036  0.005337   
2024-05-24      -0.017621      -0.001870     -0.003704 -0.011288 -0.011776   
2024-05-27      -0.009424      -0.002134      0.003717 -0.000636  0.006015   
2024-05-28       0.007032      -0.011424     -0.003419 -0.001859 -0.010980   

            TITAN.NS  ULTRACEMCO.NS  WIPRO.NS  
Date                                           
2012-01-02       NaN            NaN       NaN  
2012-01-03  0.025121       0.

In [14]:
mtl_ret=(ret_df +1).resample('M').prod() #monthly return
mtl_ret

ADANIENT.NS  ADANIPORTS.NS  APOLLOHOSP.NS  ASIANPAINT.NS  \
Date                                                                   
2012-01-31     1.440845       1.275065       1.086048       1.156353   
2012-02-29     0.901058       1.016410       0.965842       1.058508   
2012-03-31     0.813476       0.871174       1.102044       1.022536   
2012-04-30     0.921601       0.986873       0.986772       1.087847   
2012-05-31     0.927389       0.965962       1.062584       1.140283   
...                 ...            ...            ...            ...   
2024-01-31     1.102862       1.178943       1.113752       0.869342   
2024-02-29     1.045640       1.093446       0.960436       0.954037   
2024-03-31     0.973124       1.016168       1.041825       1.008806   
2024-04-30     0.955460       0.987368       0.935549       1.010240   
2024-05-31     1.061986       1.057061       0.992837       1.012274   

            AXISBANK.NS  BAJAJ-AUTO.NS  BAJAJFINSV.NS  BAJFINANCE.NS  \
Date                                                                   
2012-01-31     1.352604       1.085259       1.069397       1.244034   
2012-02-29     1.095530       1.125086       1.428873       1.053087   
2012-03-31     0.973212       0.931838       0.952715       1.030171   
2012-04-30     0.965756       0.967626       1.196566       1.073003   
2012-05-31     0.877140       0.930407       0.896132       0.972076   
...                 ...            ...            ...            ...   
2024-01-31     0.968656       1.128037       0.965298       0.936563   
2024-02-29     1.006884       1.031536       0.979414       0.946444   
2024-03-31     0.974049       1.156625       1.031403       1.115452   
2024-04-30     1.113350       0.973273       0.982450       0.955598   
2024-05-31     1.015010       1.009670       0.990495       0.992706   

            BHARTIARTL.NS   BPCL.NS  ...  SHRIRAMFIN.NS  SUNPHARMA.NS  \
Date                                 ...                                
2012-01-31       1.060615  1.190739  ...       1.387976      1.107633   
2012-02-29       0.957616  1.157476  ...       0.939189      1.003464   
2012-03-31       0.964877  1.057380  ...       1.097260      1.035520   
2012-04-30       0.918911  0.957729  ...       0.953308      1.058163   
2012-05-31       0.974396  1.040259  ...       0.920290      0.940060   
...                   ...       ...  ...            ...           ...   
2024-01-31       1.134179  1.114723  ...       1.201383      1.126246   
2024-02-29       0.959554  1.202050  ...       0.988568      1.112447   
2024-03-31       1.093693  0.997599  ...       0.967686      1.026997   
2024-04-30       1.076266  1.008217  ...       1.081320      0.926907   
2024-05-31       1.036868  1.064460  ...       0.938080      0.975235   

            TATACONSUM.NS  TATAMOTORS.NS  TATASTEEL.NS    TCS.NS  TECHM.NS  \
Date                                                                         
2012-01-31       1.184717       1.325088      1.321607  0.960801  1.140059   
2012-02-29       1.165197       1.111590      1.048248  1.079080  0.921012   
2012-03-31       0.894863       1.015870      0.998307  0.956504  1.200250   
2012-04-30       1.040498       1.150045      0.982830  1.066564  0.973639   
2012-05-31       0.920445       0.736061      0.870592  0.998997  0.959815   
...                   ...            ...           ...       ...       ...   
2024-01-31       1.028708       1.133662      0.973854  1.005945  1.047931   
2024-02-29       1.064445       1.074644      1.036043  1.073154  0.955161   
2024-03-31       0.921138       1.044833      1.106496  0.946570  0.979786   
2024-04-30       1.011084       1.015210      1.058710  0.985643  1.012339   
2024-05-31       0.988451       0.940123      1.060000  1.005038  1.040839   

            TITAN.NS  ULTRACEMCO.NS  WIPRO.NS  
Date                                           
2012-01-31  1.159007       1.052643  1.037343  
2012-02-29  1.125123       1.

In [15]:
mtl_12=mtl_ret.rolling(12).apply(np.prod).dropna()
mtl_12

ADANIENT.NS  ADANIPORTS.NS  APOLLOHOSP.NS  ASIANPAINT.NS  \
Date                                                                   
2012-12-31     0.940066       1.176983       1.424810       1.711211   
2013-01-31     0.601226       1.029402       1.347163       1.502738   
2013-02-28     0.575450       0.932728       1.426550       1.347463   
2013-03-31     0.663277       1.071042       1.313244       1.516733   
2013-04-30     0.776295       1.134585       1.323392       1.327336   
...                 ...            ...            ...            ...   
2024-01-31     1.056525       1.971191       1.492740       1.085111   
2024-02-29     2.408916       2.228880       1.386207       0.997561   
2024-03-31     1.826445       2.123516       1.474588       1.030815   
2024-04-30     1.586898       1.944665       1.317566       0.990887   
2024-05-31     1.301081       1.895513       1.277491       0.911759   

            AXISBANK.NS  BAJAJ-AUTO.NS  BAJAJFINSV.NS  BAJFINANCE.NS  \
Date                                                                   
2012-12-31     1.706781       1.444358       2.189450       2.241051   
2013-01-31     1.400074       1.265940       1.930561       1.771177   
2013-02-28     1.141286       1.095138       1.284821       1.668348   
2013-03-31     1.134793       1.071926       1.284255       1.451074   
2013-04-30     1.348480       1.157561       1.032039       1.444401   
...                 ...            ...            ...            ...   
2024-01-31     1.225046       2.008132       1.212187       1.165891   
2024-02-29     1.273664       2.160316       1.193947       1.062703   
2024-03-31     1.219802       2.354888       1.297947       1.289936   
2024-04-30     1.355698       2.008969       1.192190       1.102459   
2024-05-31     1.293545       1.968350       1.102409       0.983290   

            BHARTIARTL.NS   BPCL.NS  ...  SHRIRAMFIN.NS  SUNPHARMA.NS  \
Date                                 ...                                
2012-12-31       0.919664  1.483247  ...       1.803333      1.486773   
2013-01-31       0.928630  1.434764  ...       1.354833      1.309025   
2013-02-28       0.922044  1.123292  ...       1.270594      1.457395   
2013-03-31       0.863421  1.078758  ...       1.155972      1.436880   
2013-04-30       1.026570  1.235369  ...       1.301641      1.576936   
...                   ...       ...  ...            ...           ...   
2024-01-31       1.519797  1.463511  ...       1.917301      1.371145   
2024-02-29       1.513439  1.902789  ...       2.026509      1.649540   
2024-03-31       1.640320  1.749637  ...       1.873749      1.648408   
2024-04-30       1.654323  1.698406  ...       1.914828      1.520883   
2024-05-31       1.613190  1.778542  ...       1.710091      1.501922   

            TATACONSUM.NS  TATAMOTORS.NS  TATASTEEL.NS    TCS.NS  TECHM.NS  \
Date                                                                         
2012-12-31       1.758659       1.699647      1.256230  1.065544  1.630093   
2013-01-31       1.386079       1.222564      0.898181  1.186992  1.532132   
2013-02-28       0.997212       1.061266      0.722675  1.241458  1.750292   
2013-03-31       1.137962       0.977838      0.663169  1.348178  1.469997   
2013-04-30       1.259624       0.947718      0.652971  1.105728  1.358390   
...                   ...            ...           ...       ...       ...   
2024-01-31       1.532557       1.955762      1.135756  1.136139  1.313941   
2024-02-29       1.663708       2.258617      1.354978  1.236126  1.157782   
2024-03-31       1.546448       2.359316      1.491388  1.209114  1.132731   
2024-04-30       1.451005       2.078359      1.528485  1.186814  1.234248   
2024-05-31       1.371752       1.800399      1.653119  1.167320  1.178880   

            TITAN.NS  ULTRACEMCO.NS  WIPRO.NS  
Date                                           
2012-12-31  1.623466       1.721231  0.988722  
2013-01-31  1.376108       1.

In [17]:
top_=mtl_12.loc['2012-12-31'].nlargest(5) #finding top 5 performers
top_

BAJFINANCE.NS    2.241051
BAJAJFINSV.NS    2.189450
EICHERMOT.NS     1.943824
INDUSINDBK.NS    1.841687
SHRIRAMFIN.NS    1.803333
Name: 2012-12-31 00:00:00, dtype: float64

In [20]:
top_.name #relevant date of the top performer , in above bajaj fin has more than doubled

Timestamp('2012-12-31 00:00:00', freq='M')

In [22]:
mtl_ret[top_.name:][1:2] #this gives the next month to the top performer

ADANIENT.NS  ADANIPORTS.NS  APOLLOHOSP.NS  ASIANPAINT.NS  \
Date                                                                   
2013-01-31     0.921504       1.115185       1.026862       1.015476   

            AXISBANK.NS  BAJAJ-AUTO.NS  BAJAJFINSV.NS  BAJFINANCE.NS  \
Date                                                                   
2013-01-31     1.109543         0.9512       0.942947       0.983201   

            BHARTIARTL.NS   BPCL.NS  ...  SHRIRAMFIN.NS  SUNPHARMA.NS  \
Date                                 ...                                
2013-01-31       1.070956  1.151817  ...       1.042778      0.975212   

            TATACONSUM.NS  TATAMOTORS.NS  TATASTEEL.NS    TCS.NS  TECHM.NS  \
Date                                                                         
2013-01-31       0.933729       0.953143      0.944924  1.070311  1.071547   

            TITAN.NS  ULTRACEMCO.NS  WIPRO.NS  
Date                                           
2013-01-31  0.982416       0.957836  1.043219  

[1 rows x 50 columns]

In [23]:
relevant_ret=mtl_ret[top_.name:][1:2][top_.index]
relevant_ret

BAJFINANCE.NS  BAJAJFINSV.NS  EICHERMOT.NS  INDUSINDBK.NS  \
Date                                                                    
2013-01-31       0.983201       0.942947      0.970326       1.044959   

            SHRIRAMFIN.NS  
Date                       
2013-01-31       1.042778

In [25]:
relevant_ret.mean(axis=1)  #This is telling us that if we bought the top performer it would be 0.996842 of its value

Date
2013-01-31    0.996842
Freq: M, dtype: float64

In [27]:
def top_performers(date):
    all_=mtl_12.loc[date]
    top=all_.nlargest(5)
    relevant_ret=mtl_ret[top.name:][1:2][top.index]
    return (relevant_ret).mean(axis=1).values[0]

In [28]:
top_performers('2012-12-31')

0.9968421960832142

In [35]:
mom_ret=[]
for date in mtl_12.index[:-1]:
    mom_ret.append(top_performers(date)) 

# passing the monthly return and storing in mom_ret, skipping the last one
# becoz we dont know the future return.

In [36]:
pd.Series(mom_ret)

0      0.996842
1      0.991717
2      0.989747
3      0.997703
4      1.090375
         ...   
132    1.069835
133    1.049040
134    1.036959
135    0.996885
136    1.017264
Length: 137, dtype: float64

In [38]:
pd.Series(mom_ret).prod() #this is 11 times return

11.656284054369529

Comparing it with nifty

In [41]:
nifty=yf.download('^NSEI',start='2012-01-01')['Close']

[*********************100%%**********************]  1 of 1 completed


In [42]:
(nifty.pct_change()+1).prod()

4.803087177875052

So nifty has given 4.8 times return and our stock selection has given 11.65 times return